gensim/models/doc2vec.py

```diff
@@ -55,249 +55,560 @@
 
+        """
+        doctag_vectors, doctag_locks = self.trainables._get_doctag_trainables(doc_words)
+        doctag_indexes = [0]
+        work = zeros(self.trainables.layer1_size, dtype=REAL)
+        if not self.sg:
+            neu1 = matutils.zeros_aligned(self.trainables.layer1_size, dtype=REAL)
 
-class TaggedDocument(namedtuple('TaggedDocument', 'words tags')):
-    """
-    A single document, made up of `words` (a list of unicode string tokens)
-    and `tags` (a list of tokens). Tags may be one or more unicode string
-    tokens, but typical practice (which will also be most memory-efficient) is
-    for the tags list to include a unique integer id as the only tag.
+        for i in range(steps):
+            if self.sg:
+                train_document_dbow(
+                    self, doc_words, doctag_indexes, alpha, work,
+                    learn_words=False, learn_hidden=False, doctag_vectors=doctag_vectors, doctag_locks=doctag_locks
+                )
+            elif self.dm_concat:
+                train_document_dm_concat(
+                    self, doc_words, doctag_indexes, alpha, work, neu1,
+                    learn_words=False, learn_hidden=False, doctag_vectors=doctag_vectors, doctag_locks=doctag_locks
+                )
+            else:
+                train_document_dm(
+                    self, doc_words, doctag_indexes, alpha, work, neu1,
+                    learn_words=False, learn_hidden=False, doctag_vectors=doctag_vectors, doctag_locks=doctag_locks
+                )
+            alpha = ((alpha - min_alpha) / (steps - i)) + min_alpha
+        self._set_keyedvectors()
 
-    Replaces "sentence as a list of words" from Word2Vec.
+        return doctag_vectors[0]
 
-    """
+    def __getitem__(self, tag):
+        if isinstance(tag, string_types + integer_types + (integer,)):
+            if tag not in self.wv.vocab:
+                return self.docvecs[tag]
+            return self.wv[tag]
+        return vstack([self[i] for i in tag])
 
     def __str__(self):
-        return '%s(%s, %s)' % (self.__class__.__name__, self.words, self.tags)
+        """Abbreviated name reflecting major configuration paramaters."""
+        segments = []
+        if self.comment:
+            segments.append('"%s"' % self.comment)
+        if self.sg:
+            if self.dbow_words:
+                segments.append('dbow+w')  # also training words
+            else:
+                segments.append('dbow')  # PV-DBOW (skip-gram-style)
 
+        else:  # PV-DM...
+            if self.dm_concat:
+                segments.append('dm/c')  # ...with concatenative context layer
+            else:
+                if self.cbow_mean:
+                    segments.append('dm/m')
+                else:
+                    segments.append('dm/s')
+        segments.append('d%d' % self.trainables.vector_size)  # dimensions
+        if self.negative:
+            segments.append('n%d' % self.negative)  # negative samples
```

Comment:
Doesn't looks like a good solution (and non-oblivious what means `d`, `n`, etc).

Reply:
This is again consistent with the old code. Do you want me to change it?

Reply:
@manneshiva yes, but not in this PR I think. Need to collect all "suspicious" stuff in a list, create an issue & fix it after merging current PR.